janus_server/src/aggregator.rs

```diff
@@ -1053,12 +1057,13 @@ impl VdafOps {
         }
     }
 
-    async fn handle_aggregate_share_generic<A, E>(
+    async fn handle_aggregate_share_generic<P, A, E>(
```

Comment:
I think it makes sense to do this on `VdafOps` methods, since those are generally generic over `<A: vdaf::Aggregator>` and it makes sense there to impose a relationship between the input and output types. However I prefer to keep the methods on `Datastore` generic over `<P, A, E>` because it makes it easier to test them: in one spot, it was helpful to be able to use a different type for `aggregaton_param`.

Reply:
I tried adding the where clause `<A::AggregateShare as TryFrom<&'a [u8]>>::Error: std::fmt::Display,
` on `handle_aggregate_share_generic` instead of:
```
        E: std::fmt::Display,
        for<'a> A::AggregateShare: TryFrom<&'a [u8], Error = E>,
```

...but then I can't figure out how to then express the right type in the generic parameters to the call to`Datastore::get_batch_unit_aggregations_for_task_in_interval`. This doesn't work:

```
let batch_unit_aggregations = tx
                        .get_batch_unit_aggregations_for_task_in_interval::<
                            _,
                            A::AggregateShare,
                            for<'a> <A::AggregateShare as TryFrom<&'a [u8]>>::Error,
                        >(
                            task.id,
                            aggregate_share_req.batch_interval,
                            &aggregation_param,
                        )
                        .await?;
```
I can't figure out where to place `for<'a>` to make this work. Anyway, that's an interesting trait golf exercise, but this works fine if we introduce the type `E` and bind it to `A::AggregateShare` in the method's `where` block so I'll call that good enough.

Reply:
I was able to get it to work by removing E as a type parameter entirely. https://github.com/divviup/janus/commit/56846927aeb2369bbf98d091e33db96de080f159 I'm not sold on whether this is better or worse, as it needs more quantifier constraints. There's also some give and take with type inference in the tests.

Generally speaking, I don't think putting a quantifier at that position in a method's turbofish types would work, as each monomorphization needs one concrete type, rather than a class of types with varying lifetime parameters.